Sidecar frontend readme docs

jaredcwhite opened the pull request:
<!-- See https://github.com/github/view_component/blob/master/CONTRIBUTING.md#submitting-a-pull-request  -->

### Summary

This is a first draft of adding info on sidecar frontend assets to the README (related to #168). I need to test out the examples more to make sure they're bug-free and some of the wording could be streamlined, but I wanted your overall feedback before tweaking it too much.

### Other Information

I feel like this might be a good candidate for adding to a separate .md docs file somewhere, as I'm not quite sure it integrates holistically into the existing README flow. Let me know what you think.

joelhawksley commented on README.md:
```diff
@@ -321,6 +321,129 @@ end
+
+An interesting design pattern the community is exploring is including Javascript files and possibly stylesheets alongside components, which allows backend and frontend development of a component to proceed in parallel. These additional frontend assets are sometimes called "sidecar" assets or files.
+
+To use the Rails Webpacker gem to compile sidecar assets located in `app/components`, you will need to add a couple of setup steps.
```
```suggestion
To use the Webpacker gem to compile sidecar assets located in `app/components`:
```

joelhawksley commented on README.md:
```diff
@@ -321,6 +321,129 @@ end
+
+To use the Rails Webpacker gem to compile sidecar assets located in `app/components`, you will need to add a couple of setup steps.
+
+1. In `config/webpacker.yml`, you will need to add `"app/components"` to the `resolved_paths` array (e.g. `resolved_paths: ["app/components"]`).
```
```suggestion
1. In `config/webpacker.yml`,  add `"app/components"` to the `resolved_paths` array (e.g. `resolved_paths: ["app/components"]`).
```

joelhawksley commented on README.md:
```diff
@@ -321,6 +321,129 @@ end
+To use the Rails Webpacker gem to compile sidecar assets located in `app/components`, you will need to add a couple of setup steps.
+
+1. In `config/webpacker.yml`, you will need to add `"app/components"` to the `resolved_paths` array (e.g. `resolved_paths: ["app/components"]`).
+2. In your Webpack entry file (likely `app/javascript/packs/application.js`), you'll need to add an import statement to a helper file and in the helper file import all the components' Javascript.
```
```suggestion
2. In the Webpack entry file (often `app/javascript/packs/application.js`), add an import statement to a helper file, and in the helper file, import the components' Javascript:
```

joelhawksley commented on README.md:
````diff
@@ -321,6 +321,129 @@ end
+importAll(require.context("../components", true, /_component.js$/))
+```
+
+Now any file with that suffix, for example `app/components/widget_component.js` will get compiled into the Webpack bundle. If that file itself imports another file, for example `app/components/widget_component.css`, that will also get compiled and bundled into Webpack's output stylesheet if you're using Webpack for styles.
````
```suggestion
Any file with the `_component.js` suffix, for example `app/components/widget_component.js`, will get compiled into the Webpack bundle. If that file itself imports another file, for example `app/components/widget_component.css`, that will also get compiled and bundled into Webpack's output stylesheet if Webpack is being used for styles.
```

joelhawksley commented on README.md:
```diff
@@ -321,6 +321,129 @@ end
+
+Now any file with that suffix, for example `app/components/widget_component.js` will get compiled into the Webpack bundle. If that file itself imports another file, for example `app/components/widget_component.css`, that will also get compiled and bundled into Webpack's output stylesheet if you're using Webpack for styles.
+
+#### Scoping Frontend Component Code
```
```suggestion
#### Encapsulating sidecar assets
```

joelhawksley commented on README.md:
```diff
@@ -321,6 +321,129 @@ end
+
+#### Scoping Frontend Component Code
+
+There are many approaches to this, the goal being that any Javascript or CSS doesn't "leak" out into other parts of the application.
```
```suggestion
Ideally, sidecar Javascript/CSS should not "leak" out of the context of its associated component. 
```

joelhawksley commented on README.md:
```diff
@@ -321,6 +321,129 @@ end
+
+There are many approaches to this, the goal being that any Javascript or CSS doesn't "leak" out into other parts of the application.
+
+One approach is to use standards-based Web Components. Because all Javascript functionality, internal markup, and styles are contained within the shadow root of the Web Component, it minimizes the risk of unexpected side-effects regarding other components.
```
```suggestion
One approach is to use Web Components, which contain all Javascript functionality, internal markup, and styles within the shadow root of the Web Component.
```

joelhawksley commented on README.md:
```diff
@@ -321,6 +321,129 @@ end
+
+One approach is to use standards-based Web Components. Because all Javascript functionality, internal markup, and styles are contained within the shadow root of the Web Component, it minimizes the risk of unexpected side-effects regarding other components.
+
+Here's an example of how that would work. You'd define three files per component: the `ViewComponent` Ruby object, the Rails view template, and the corresponding Javascript implementation of the Web Component.
```
```suggestion
For example:
```

joelhawksley commented on README.md:
````diff
@@ -321,6 +321,129 @@ end
+
+`app/components/comment_component.rb`
+```ruby
+class CommentComponent < ApplicationComponent
````
```suggestion
class CommentComponent < ViewComponent::Base
```

joelhawksley commented on README.md:
```diff
@@ -321,6 +321,129 @@ end
+#### Scoping Frontend Component Code
+
+There are many approaches to this, the goal being that any Javascript or CSS doesn't "leak" out into other parts of the application.
+
```
```suggestion

##### Web Components
```

joelhawksley commented on README.md:
````diff
@@ -321,6 +321,129 @@ end
+}
+customElements.define('my-comment', Comment)
+```
+
````
```suggestion

##### Stimulus
```

joelhawksley commented on README.md:
````diff
@@ -321,6 +321,129 @@ end
+customElements.define('my-comment', Comment)
+```
+
+Another approach is to use a framework like Stimulus and create a 1:1 mapping between a Stimulus controller and a component. We'll leave the controller and HTML markup as an exercise for the reader, but in order to load in Stimulus controllers from the `app/components` tree, you'll have to amend the Stimulus boot code in `app/javascript/packs/application.js` as follows:
````
```suggestion
In Stimulus, create a 1:1 mapping between a Stimulus controller and a component. In order to load in Stimulus controllers from the `app/components` tree, amend the Stimulus boot code in `app/javascript/packs/application.js`:
```

joelhawksley commented on README.md:
````diff
@@ -321,6 +321,129 @@ end
+)
+```
+
+This will allow you to create files such as `app/components/widget_controller.js` and as long as the controller identifier matches the `data-controller` attribute in the component's HTML template, you're all set.
````
```suggestion
This will allow you to create files such as `app/components/widget_controller.js`, where the controller identifier matches the `data-controller` attribute in the component's HTML template.
```

joelhawksley commented:
Also, can you add yourself to the README?

jaredcwhite commented:
@joelhawksley I appreciate the detailed response! I'll review your suggestions and provide an update by tomorrow.

jaredcwhite commented:
@joelhawksley Hey, so I incorporated your copy recommendations, fixed a few bugs in the example code, and added myself to the bottom of the README. Let me know what you think!

joelhawksley commented:
Thank you for your contribution @jaredcwhite!

rmacklin commented on README.md:
```diff
@@ -493,10 +624,10 @@ Bug reports and pull requests are welcome on GitHub at https://github.com/github
-|:---:|:---:|:---:|:---:|
-|@blakewilliams|@seanpdoyle|@tclem|@nashby|
-|Boston, MA|New York, NY|San Francisco, CA|Minsk|
+|<img src="https://avatars.githubusercontent.com/blakewilliams?s=256" alt="blakewilliams" width="128" />|<img src="https://avatars.githubusercontent.com/seanpdoyle?s=256" alt="seanpdoyle" width="128" />|<img src="https://avatars.githubusercontent.com/tclem?s=256" alt="tclem" width="128" />|<img src="https://avatars.githubusercontent.com/nashby?s=256" alt="nashby" width="128" />|<img src="https:// ...
```
Minor, but I think the `alt` tag should be `jaredcwhite`
